### DISCUSSION TITLE
Custom SVG to PNG

### DISCUSSION THREAD

**Discussion by nidhintrell** [chillerlan/php-qrcode#216] [Q&A]
Hi, thanks for this library. 
I am trying to achieve custom QR Code using this library which includes patterns, centre logo and colours. All this is good and i am able to achieve it. Now i need a way to save it as PNG in my system. Not sure how to achieve that. Tried few stuffs but path conversion is not happening as it was supposed too. How can i convert the SVG to PNG or is there a way that i can direclty save the QR as png.

**Comment by codemasher** [comment]
Hi! There's actually 2 ways, the first would be [using ImageMagick](https://stackoverflow.com/questions/29994680/how-to-convert-svg-to-png-using-imagemagick-in-php), however this may not necessarily produce the desired results, especially when using external styling via `<defs>` or CSS attributes. The other way would be to use [the command line interface of Inkscape](https://wiki.inkscape.org/wiki/Using_the_Command_Line), which you can call from inside PHP via [`exec()`](https://www.php.net/manual/en/function.exec). The example images from the readme were rendered with Inkscape:

![example png](https://github.com/chillerlan/php-qrcode/blob/main/.github/images/example_svg.png?raw=true)

**Reply by nidhintrell** [reply]
@codemasher thanks for this. yes tried both. Inkscape works.but not a fan of running exec() from web interface unless some kind of cron.. was chking if there is any other method to achieve the same within the library.

**Reply by codemasher** [reply]
In that case you will have to stick with ImageMagick and its quirks then.

**Reply by nidhintrell** [reply]
@codemasher ya saw the imge gets ruined with that .. no issues will go with exec()

**Reply by codemasher** [reply]
May i ask what your use case is? You say "web interface" but want it as PNG anyway?

**Reply by codemasher** [reply]
Update: i tried a render through ImageMagick of the SVG example and it looks surprisingly good, however, you might need to specify the `width` and `height` attributes [in the SVG header over here](https://github.com/chillerlan/php-qrcode/blob/f88b53adf28a93da017d21596ecc4bb9c048f736/src/Output/QRMarkupSVG.php#L91-L111). 

```php
$im = new Imagick;
$im->readImageBlob($svg);
$im->setImageFormat('png32');

header('Content-type: image/png');

echo $im->getImageBlob();
```

This is the result directly from ImageMagick:

![imagickConvertSVGtoPNG](https://github.com/chillerlan/php-qrcode/assets/592497/0f1fba5e-a8df-41b7-b183-436a1e24ffdc)

**Reply by nidhintrell** [reply]
Yes i can use the SVG for web and App but i'm giving an option to export as an image

**Comment by codemasher** [comment]
There, i made a quick example: https://github.com/chillerlan/php-qrcode/commit/132520b0d0e9314d108eeb6459fada0aa811b1eb

**Reply by nidhintrell** [reply]
Let me try this, thanks

**Comment by nidhintrell** [comment]
@codemasher i have an image at the centre of the qr. im using svg fr center image.. when using that the conversion fails in your example. example svg used
![qrcode_test](https://github.com/chillerlan/php-qrcode/assets/67859556/2099cb38-37fd-4ed4-840b-3c50d02ca2b7)

**Reply by codemasher** [reply]
Yea, this is why i said ImageMagick is unreliable to just convert SVG.
But then again since you're in a web app, you can just [export the SVG via javascript](https://stackoverflow.com/a/74026755). (see also https://github.com/chillerlan/js-qrcode)

**Reply by nidhintrell** [reply]
@codemasher Most of it is in App. Looks like will use exec method

**Comment by nidhintrell** [comment]
@codemasher  wanted to know how can i give color to the 3 corners?

**Reply by codemasher** [reply]
In your custom created SVG? Via CSS in the `<defs>` in the same way i added the gradient in the [example](https://github.com/chillerlan/php-qrcode/blob/main/examples/svg.php).

**Reply by codemasher** [reply]
[Here's a more complex example](https://github.com/chillerlan/php-qrcode/blob/main/examples/svgRoundQuietzone.php) that shows how to color different layers, which is what you want.

**Reply by nidhintrell** [reply]
yes

**Reply by semaf** [reply]
What about at GD or Imagick? Didnt find anything in the examples.

**Reply by codemasher** [reply]
The [`collectModules()` method](https://github.com/chillerlan/php-qrcode/blob/a301af5b7faa7fcd690b29f38b283616a7d93b89/examples/svgRoundQuietzone.php#L176-L216) for the markup based output classes is just a more elaborate version of the [`drawImage()` method](https://github.com/chillerlan/php-qrcode/blob/main/src/Output/QRImagick.php#L194-L205) in both, GD and Imagick. You can use it as a basis for your own implementations.

**Comment by nidhintrell** [comment]
@codemasher been referring this one https://github.com/chillerlan/php-qrcode/discussions/150. but here i dont see the create circle method so using id  circle with css  doesnt work out. additionally how to make it transaprent background in the same?

**Reply by codemasher** [reply]
The example in #150 just adds the path for the finder patterns to the path element of the qr code data - you would need to put it in a separate path instead.
